Add manage_containers.py

Willenst opened the pull request:
Hello, I've decided to make a python script for these issues #12 #13 

So, here is python script, documentation, tests, and some additions for old clangs support in Dockerfile and make_linux.py!

a13xp0p0v commented:
Hello @Willenst,

Thanks for your work!

Please see some notes below.

a13xp0p0v commented on README.md:
```diff
@@ -44,6 +44,15 @@ __Supported clang versions:__
 
+__Additional supported clang versions (only for manage_containers.py):__
+ - clang-5.0
+ - clang-6.0
```
It would be nice to call these containers `clang-5` and `clang-6`

a13xp0p0v commented on manage_containers.py:
```diff
@@ -0,0 +1,189 @@
+        list_containers(containers)
+
+    if args.add:
+        add_handler(args.add, containers)
```
It would match `list_containers`, by the way.

a13xp0p0v commented on manage_containers.py:
```diff
@@ -0,0 +1,189 @@
+    """Print built containers"""
+    for c in containers:
+        if c.id:
+            print(f'Ubuntu-{c.ubuntu} with gcc-{c.gcc} and clang-{c.clang}: [+]')
```
I would use the output field width here to make it pretty.

a13xp0p0v commented on manage_containers.py:
```diff
@@ -0,0 +1,189 @@
+    for c in containers:
+        if 'gcc-' + c.gcc == needed_compiler:
+            if c.id:
+                sys.exit(f'[!] ERROR: container with {needed_compiler} already exists!')
```
I think we should print `WARNING` here, like we do above.

a13xp0p0v commented on manage_containers.py:
```diff
@@ -0,0 +1,189 @@
+    print('We need to use sudo for running docker')
+    return 'sudo'
+
+def add_handler(needed_compiler, containers):
```
This function contains a lot of copy-pasted code.
Please refactor it. You can make it much simpler.

a13xp0p0v reviewed (changes requested):
These comments are mostly about the code style.

Please fix them and rebase the branch onto the fresh master.

And then we will test the functionality carefully.

Thanks again for your work! 👍

a13xp0p0v commented on manage_containers.py:
```diff
@@ -0,0 +1,189 @@
+        id (str): Docker image id
+    """
+
+    sudo = ''
```
I see you implemented the `sudo_wrapper`. Ok!

a13xp0p0v commented on manage_containers.py:
```diff
@@ -0,0 +1,179 @@
+        return self.id
+
+def check_group():
+    """Check if the user is in the Docker group, return 'sudo' if not"""
```
How about making this a part of the class constructor?

Initially, you can have `sudo_wrapper = None`.

Then in the `__init__()` you can initialize it to `''` or `'sudo'`.

Willenst replied:
Not sure if it's a great idea because we need to check whether sudo_wrapper is None each time a container instance is created. Anyway, I've made it, so you can check whether this is okay:

https://github.com/a13xp0p0v/kernel-build-containers/pull/22/commits/77a71b90776ce3d9e646d87cf2f97d0d76c14a02

a13xp0p0v commented on README.md:
````diff
@@ -75,6 +82,41 @@ kernel-build-container   gcc-5      ed6270a767c6   8 hours ago          720MB
+### Building all containers (manage_containers.py compact alternative):
+
+```console
+python3 manage_containers.py -a all -q
````
Also, please add other usage examples of this tool to the README.
Thanks!

Willenst replied:
Expanded a bit :+1: 

https://github.com/a13xp0p0v/kernel-build-containers/pull/22/commits/3cae5e9e421255fdce98316e0ed463609fb271b5

a13xp0p0v commented on manage_containers.py:
```diff
@@ -0,0 +1,179 @@
+        """Remove the Docker container if it exists and not running"""
+        running = subprocess.run(f'{Container.sudo_wrapper} docker ps | '
+                                 f'grep -E "kernel-build-container:'
+                                 f'(gcc-{self.gcc}|clang-{self.clang})" || true',
```
It's better to use a list instead of a large string here.

a13xp0p0v replied:
Agree.
Maybe it's better to parse the results of `docker ps` here in python script to avoid this long command.

Willenst replied:
better now?

https://github.com/a13xp0p0v/kernel-build-containers/pull/22/commits/76a512ca88e5b286ac4240aa94c6a0b26cae04aa

a13xp0p0v commented on manage_containers.py:
```diff
@@ -0,0 +1,189 @@
+
+    def rm(self):
+        """Removes the Docker container if it exists and not running"""
+        running = subprocess.run(f"{Container.sudo} docker ps | "
```
This is still relevant for other cases.

Willenst replied:
My bad, I missed some. Now, double quotations are only in the comments and one subprocess block, which I think needs to be refactored first 

https://github.com/a13xp0p0v/kernel-build-containers/pull/22/commits/be0388195dc689bfbf4b0c6261c53952a766fcf0.

Willenst replied:
So, because of this refactor:

https://github.com/a13xp0p0v/kernel-build-containers/pull/22/commits/76a512ca88e5b286ac4240aa94c6a0b26cae04aa

there are no more double quotation marks left (except comments).

Willenst commented:
Re-based on current PR, answered for all conversations, waiting for next review!

Willenst commented on manage_containers.py:
```diff
@@ -0,0 +1,189 @@
+
+    if args.list:
+        list_containers(containers)
+
```
https://github.com/a13xp0p0v/kernel-build-containers/pull/22/commits/0c30a161f74f53d94d73b7ba316acefe5b8a637c

a13xp0p0v commented:
Hello @Willenst,

Unfortunately, this branch is broken in multiple places.

1) Please check it once again after the commit https://github.com/a13xp0p0v/kernel-build-containers/pull/22/commits/9b1c4b9e93ca608fd5b39379d89462bc0ad4e142

There are things like this:
```
  File "./manage_containers.py", line 124
    print(f'\n{'Ubuntu':<6} | {'GCC':<6} | {'Clang':<6} | {'Status':<6}')
                ^
SyntaxError: invalid syntax
```

It can be fixed this way:
```
 def list_containers(containers):
     """Print built containers"""
-    print(f'\n{'Ubuntu':<6} | {'GCC':<6} | {'Clang':<6} | {'Status':<6}')
+    print(f'\n{"Ubuntu":<6} | {"GCC":<6} | {"Clang":<6} | {"Status":<6}')
```

2) An empty sudo wrapper is broken (if the current user is in the `docker` group):
```
$ ./manage_containers.py -a gcc-10
Adding Ubuntu-20.04 container with GCC-10 and Clang-11
Traceback (most recent call last):
  File "./manage_containers.py", line 183, in <module>
    main()
  File "./manage_containers.py", line 173, in main
    add_containers(args.add, containers)
  File "./manage_containers.py", line 109, in add_containers
    c.add()
  File "./manage_containers.py", line 61, in add
    subprocess.run([self.sudo_wrapper, 'docker', 'build', *build_args, '.'],
  File "/usr/lib/python3.8/subprocess.py", line 493, in run
    with Popen(*popenargs, **kwargs) as process:
  File "/usr/lib/python3.8/subprocess.py", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/usr/lib/python3.8/subprocess.py", line 1704, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
PermissionError: [Errno 13] Permission denied: ''
```

Also, could you please reorganize the commits in this branch after fixing those bugs.

Thanks you so much for your work!

Willenst commented:
Hello @a13xp0p0v, thanks a lot for guiding me through this PR! It seems like the reason for the first crash was old Python. I adapted the code for it and also fixed it to work properly for the `docker` group. Additionally, I added some checks into the tests! I'm waiting for your review to start squashing commits!

a13xp0p0v commented:
@Willenst, nice. I think it's time to rearrange the commits.
Thanks!
